src/Resources/contao/library/Contao/User.php

```diff
@@ -258,481 +282,466 @@ public function getData()
 		{
-			$this->locked = $time;
-			$this->loginCount = \Config::get('loginCount');
-			$this->save();
+			return true;
+		}
 
-			// Add a log entry and the error message, because checkAccountStatus() will not be called (see #4444)
-			$this->log('User "' . $this->username . '" has been locked for ' . ceil(\Config::get('lockPeriod') / 60) . ' minutes', __METHOD__, TL_ACCESS);
-			\Message::addError(sprintf($GLOBALS['TL_LANG']['ERR']['accountLocked'], ceil((($this->locked + \Config::get('lockPeriod')) - $time) / 60)));
+		return false;
+	}
 
-			// Send admin notification
-			if (\Config::get('adminEmail') != '')
-			{
-				$objEmail = new \Email();
-				$objEmail->subject = $GLOBALS['TL_LANG']['MSC']['lockedAccount'][0];
-				$objEmail->text = sprintf($GLOBALS['TL_LANG']['MSC']['lockedAccount'][1], $this->username, ((TL_MODE == 'FE') ? $this->firstname . " " . $this->lastname : $this->name), \Idna::decode(\Environment::get('base')), ceil(\Config::get('lockPeriod') / 60));
-				$objEmail->sendTo(\Config::get('adminEmail'));
-			}
 
-			return false;
-		}
+	/**
+	 * Set all user properties from a database record
+	 */
+	abstract protected function setUserFromDb();
 
-		// Check the account status
-		if ($this->checkAccountStatus() == false)
-		{
-			return false;
-		}
 
-		$blnNeedsRehash = true;
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getRoles()
+	{
+		return array();
+	}
 
-		// Handle old sha1() passwords with an optional salt
-		if (preg_match('/^[a-f0-9]{40}(:[a-f0-9]{23})?$/', $this->password))
-		{
-			list($strPassword, $strSalt) = explode(':', $this->password);
-			$blnAuthenticated = ($strPassword === sha1($strSalt . $request->request->get('password')));
-		}
-		else
-		{
-			$blnAuthenticated = password_verify($request->request->get('password'), $this->password);
-			$blnNeedsRehash = password_needs_rehash($this->password, PASSWORD_DEFAULT);
-		}
 
-		// Re-hash the password if the algorithm has changed
-		if ($blnAuthenticated && $blnNeedsRehash)
-		{
-			$this->password = password_hash($request->request->get('password'), PASSWORD_DEFAULT);
-		}
+	/**
+	 * {@inheritdoc}
+	 */
+	public static function loadUserByUsername($username)
+	{
+		/** @var EventDispatcherInterface $eventDispatcher */
+		$eventDispatcher = \System::getContainer()->get('event_dispatcher');
+
+		/** @var Request $request */
+		$request = \System::getContainer()->get('request_stack')->getCurrentRequest();
+		$password = $request->request->get('password');
+
+		$user = new static();
 
-		// HOOK: pass credentials to callback functions
-		if (!$blnAuthenticated && isset($GLOBALS['TL_HOOKS']['checkCredentials']) && \is_array($GLOBALS['TL_HOOKS']['checkCredentials']))
+		// Load the user object
+		if ($user->findBy('username', $username) === false)
 		{
-			foreach ($GLOBALS['TL_HOOKS']['checkCredentials'] as $callback)
+			// Return if its not a real login attempt
+			if ($password === null && !$request->isMethod(Request::METHOD_POST))
```

Comment:
What's the purpose of this line? What do you mean by "not a real login attempt"?

Reply:
`loadUserByUsername` will be called even on `refreshUser`, but the following logic only needs to be executed on a real login and not on a refresh.